[quant] Input-Weight Equalization - tests

angelayi opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #60883
* #60653
* #60779
* **#60378**
* #60272
* #60034
* #59964

Summary: Created the following unit-tests to check that our equalization algorithm is as expected:
- Check the equalization scales calculated and stored in the graph are as expected
- Check the scaled weights and biases are as expected
- Check that the min/max values in the quantization observers are as expected
- Check that the graphs with equalization are structured in the same way as graphs without equalization (except that equalized graphs have additional equalization scale and mul nodes) before and after quantization

Test Plan: `python test/test_quantization TestEqualizeFx.test_input_weight_equalization_equalization_scales`
`python test/test_quantization TestEqualizeFx.test_input_weight_equalization_weights_bias`
`python test/test_quantization TestEqualizeFx.test_input_activation_values`
`python test/test_quantization TestEqualizeFx.test_input_weight_equalization_graphs`

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29406942](https://our.internmc.facebook.com/intern/diff/D29406942)

angelayi committed:
[quant] Input-Weight Equalization - tests

Summary: Created the following unit-tests to check that our equalization
algorithm is accurate:
- Check the equalization scales calculated and stored in the graph are as expected
- Check the scaled weights and biases are as expected
- Check that the min/max values in the quantization observers are as
expected
- Check that the graphs with equalization are structured in the same way
as graphs without equalization (except that equalized graphs have
additional equalization scale and mul nodes) before and after
quantization

Test Plan: `python test/test_quantization TestEqualizeFx.test_input_weight_equalization_equalization_scales`
`python test/test_quantization TestEqualizeFx.test_input_weight_equalization_weights_bias`
`python test/test_quantization TestEqualizeFx.test_input_activation_values`
`python test/test_quantization TestEqualizeFx.test_input_weight_equalization_graphs`

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

angelayi committed:
Update on "[quant] Input-Weight Equalization - tests"


Summary: Created the following unit-tests to check that our equalization algorithm is as expected:
- Check the equalization scales calculated and stored in the graph are as expected
- Check the scaled weights and biases are as expected
- Check that the min/max values in the quantization observers are as expected
- Check that the graphs with equalization are structured in the same way as graphs without equalization (except that equalized graphs have additional equalization scale and mul nodes) before and after quantization

Modifications to the equalization functions:
- Regardless of if the input equalization observer will be deleted, we will still update the quantization observer's min/max values so that the qparams that get calculated with reflect the scaled values. Originally, we were only updating the quantization's observer's min/max values when it was not between connecting linear layers. However, this would mean that the quantization observers between two connected linear layers would not have the correct qparams.
- For functional layers, before deleting the weight equalization observer nodes, we will clear the min/max values of the weight quantization observer. If we do not clear the values, then the weight quantization observers will compare the scaled min/max weight values against the unscaled min/max weight values.

Test Plan: `python test/test_quantization TestEqualizeFx.test_input_weight_equalization_equalization_scales`
`python test/test_quantization TestEqualizeFx.test_input_weight_equalization_weights_bias`
`python test/test_quantization TestEqualizeFx.test_input_activation_values`
`python test/test_quantization TestEqualizeFx.test_input_weight_equalization_graphs`

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

angelayi committed:
Update on "[quant] Input-Weight Equalization - tests"


Summary: Created the following unit-tests to check that our equalization algorithm is as expected:
- Check the equalization scales calculated and stored in the graph are as expected
- Check the scaled weights and biases are as expected
- Check that the min/max values in the quantization observers are as expected
- Check that the graphs with equalization are structured in the same way as graphs without equalization (except that equalized graphs have additional equalization scale and mul nodes) before and after quantization

Modifications to the equalization functions:
- Regardless of if the input equalization observer will be deleted, we will still update the quantization observer's min/max values so that the qparams that get calculated with reflect the scaled values. Originally, we were only updating the quantization's observer's min/max values when it was not between connecting linear layers. However, this would mean that the quantization observers between two connected linear layers would not have the correct qparams.
- For functional layers, before deleting the weight equalization observer nodes, we will clear the min/max values of the weight quantization observer. If we do not clear the values, then the weight quantization observers will compare the scaled min/max weight values against the unscaled min/max weight values.

Test Plan: `python test/test_quantization TestEqualizeFx.test_input_weight_equalization_equalization_scales`
`python test/test_quantization TestEqualizeFx.test_input_weight_equalization_weights_bias`
`python test/test_quantization TestEqualizeFx.test_input_activation_values`
`python test/test_quantization TestEqualizeFx.test_input_weight_equalization_graphs`

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

angelayi committed:
Update on "[quant] Input-Weight Equalization - tests"


Summary: Created the following unit-tests to check that our equalization algorithm is as expected:
- Check the equalization scales calculated and stored in the graph are as expected
- Check the scaled weights and biases are as expected
- Check that the min/max values in the quantization observers are as expected
- Check that the graphs with equalization are structured in the same way as graphs without equalization (except that equalized graphs have additional equalization scale and mul nodes) before and after quantization

Modifications to the equalization functions:
- Regardless of if the input equalization observer will be deleted, we will still update the quantization observer's min/max values so that the qparams that get calculated with reflect the scaled values. Originally, we were only updating the quantization's observer's min/max values when it was not between connecting linear layers. However, this would mean that the quantization observers between two connected linear layers would not have the correct qparams.
- For functional layers, before deleting the weight equalization observer nodes, we will clear the min/max values of the weight quantization observer. If we do not clear the values, then the weight quantization observers will compare the scaled min/max weight values against the unscaled min/max weight values.

Test Plan: `python test/test_quantization TestEqualizeFx.test_input_weight_equalization_equalization_scales`
`python test/test_quantization TestEqualizeFx.test_input_weight_equalization_weights_bias`
`python test/test_quantization TestEqualizeFx.test_input_activation_values`
`python test/test_quantization TestEqualizeFx.test_input_weight_equalization_graphs`

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

angelayi committed:
Update on "[quant] Input-Weight Equalization - tests"


Summary: Created the following unit-tests to check that our equalization algorithm is as expected:
- Check the equalization scales calculated and stored in the graph are as expected
- Check the scaled weights and biases are as expected
- Check that the min/max values in the quantization observers are as expected
- Check that the graphs with equalization are structured in the same way as graphs without equalization (except that equalized graphs have additional equalization scale and mul nodes) before and after quantization

Modifications to the equalization functions:
- Regardless of if the input equalization observer will be deleted, we will still update the quantization observer's min/max values so that the qparams that get calculated with reflect the scaled values. Originally, we were only updating the quantization's observer's min/max values when it was not between connecting linear layers. However, this would mean that the quantization observers between two connected linear layers would not have the correct qparams.
- For functional layers, before deleting the weight equalization observer nodes, we will clear the min/max values of the weight quantization observer. If we do not clear the values, then the weight quantization observers will compare the scaled min/max weight values against the unscaled min/max weight values.

Test Plan: `python test/test_quantization TestEqualizeFx.test_input_weight_equalization_equalization_scales`
`python test/test_quantization TestEqualizeFx.test_input_weight_equalization_weights_bias`
`python test/test_quantization TestEqualizeFx.test_input_activation_values`
`python test/test_quantization TestEqualizeFx.test_input_weight_equalization_graphs`

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

angelayi committed:
Update on "[quant] Input-Weight Equalization - tests"


Summary: Created the following unit-tests to check that our equalization algorithm is as expected:
- Check the equalization scales calculated and stored in the graph are as expected
- Check the scaled weights and biases are as expected
- Check that the min/max values in the quantization observers are as expected
- Check that the graphs with equalization are structured in the same way as graphs without equalization (except that equalized graphs have additional equalization scale and mul nodes) before and after quantization

Test Plan: `python test/test_quantization TestEqualizeFx.test_input_weight_equalization_equalization_scales`
`python test/test_quantization TestEqualizeFx.test_input_weight_equalization_weights_bias`
`python test/test_quantization TestEqualizeFx.test_input_activation_values`
`python test/test_quantization TestEqualizeFx.test_input_weight_equalization_graphs`

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

supriyar commented on test/quantization/fx/test_equalize_fx.py:
```diff
@@ -11,17 +12,91 @@
+
+    def forward(self, x):
+        x = self.linear1(x)
+        x = torch.add(x, 5)
```
Is it intended to have an add here since the model name says Linear2FP32Module

angelayi replied:
Sorry, I think the comment above is incorrect. I meant to say that it has 2 linear layers with an add operation in between. Would the name `LinearAddModule` be better?

supriyar commented on test/quantization/fx/test_equalize_fx.py:
```diff
@@ -135,220 +210,296 @@ def test_input_weight_eq_observer(self, input_qdtype, input_qscheme, weight_qdty
+            eq_scale = self.calculate_equalization_scale_ref(x, weight)
+            exp_eq_scales.append(eq_scale)
+
+            x = x @ weight.T + bias
```
This would work only for single/connected linear modules, right? Should we make that clearer?

angelayi replied:
Yes, I can leave a comment on the functions.

supriyar commented on test/quantization/fx/test_equalize_fx.py:
```diff
@@ -135,220 +210,296 @@ def test_input_weight_eq_observer(self, input_qdtype, input_qscheme, weight_qdty
+
+        return True
+
+    def test_input_weight_equalization_graphs(self):
```
Can we use `checkGraphModeFxOp` with an expected node_list and node_occurance? Does testing it this way provide any additional coverage?

angelayi replied:
Yes, I think I could use `checkGraphModeFxOp` with the `expected_node_list` which checks the orders of the occurrences of the nodes - this way we can accurately check the placement of the mul node.

angelayi replied:
I misread the `checkGraphModeFxOp` function and I don't think I'll be able to use it to check the order of occurrences of the node because it runs `prepare_fx` and `convert_fx` without equalization on the model that gets passed in. But, I found that `checkGraphModuleNodes` with an `expected_node_list` also works the same (I have updated the code).

angelayi committed:
Update on "[quant] Input-Weight Equalization - tests"


Summary: Created the following unit-tests to check that our equalization algorithm is as expected:
- Check the equalization scales calculated and stored in the graph are as expected
- Check the scaled weights and biases are as expected
- Check that the min/max values in the quantization observers are as expected
- Check that the graphs with equalization are structured in the same way as graphs without equalization (except that equalized graphs have additional equalization scale and mul nodes) before and after quantization

Test Plan: `python test/test_quantization TestEqualizeFx.test_input_weight_equalization_equalization_scales`
`python test/test_quantization TestEqualizeFx.test_input_weight_equalization_weights_bias`
`python test/test_quantization TestEqualizeFx.test_input_activation_values`
`python test/test_quantization TestEqualizeFx.test_input_weight_equalization_graphs`

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

angelayi committed:
Update on "[quant] Input-Weight Equalization - tests"


Summary: Created the following unit-tests to check that our equalization algorithm is as expected:
- Check the equalization scales calculated and stored in the graph are as expected
- Check the scaled weights and biases are as expected
- Check that the min/max values in the quantization observers are as expected
- Check that the graphs with equalization are structured in the same way as graphs without equalization (except that equalized graphs have additional equalization scale and mul nodes) before and after quantization

Test Plan: `python test/test_quantization TestEqualizeFx.test_input_weight_equalization_equalization_scales`
`python test/test_quantization TestEqualizeFx.test_input_weight_equalization_weights_bias`
`python test/test_quantization TestEqualizeFx.test_input_activation_values`
`python test/test_quantization TestEqualizeFx.test_input_weight_equalization_graphs`

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 commented on test/quantization/fx/test_equalize_fx.py:
```diff
@@ -11,17 +13,92 @@
 
+# Test where we have a nn.Linear layer followed by a fp32 operation (add layer
+# without a qconfig) followed by another nn.Linear layer
+class LinearAddModule(nn.Module):
```
nit: feel free to add these to common_quantization.py as well

jerryzh168 commented on test/quantization/fx/test_equalize_fx.py:
```diff
@@ -11,17 +13,92 @@
+        x = self.linear2(x)
+        return x
+
+class Linear(torch.nn.Module):
```
maybe rename this since we have torch.nn.Linear as well

supriyar commented on test/quantization/fx/test_equalize_fx.py:
```diff
@@ -135,220 +212,360 @@ def test_input_weight_eq_observer(self, input_qdtype, input_qscheme, weight_qdty
+            equalized_model = _convert_equalization_ref(prepared)
+
+            # Check that the graph before quantization is pretty much the same
+            self.assertTrue(self.check_orig_and_eq_graphs(prepared_model, equalized_model))
```
Do we still need this function to check if we are using `checkGraphModuleNodes`?

angelayi replied:
I guess not, I can remove that then!

angelayi committed:
Update on "[quant] Input-Weight Equalization - tests"


Summary: Created the following unit-tests to check that our equalization algorithm is as expected:
- Check the equalization scales calculated and stored in the graph are as expected
- Check the scaled weights and biases are as expected
- Check that the min/max values in the quantization observers are as expected
- Check that the graphs with equalization are structured in the same way as graphs without equalization (except that equalized graphs have additional equalization scale and mul nodes) before and after quantization

Test Plan: `python test/test_quantization TestEqualizeFx.test_input_weight_equalization_equalization_scales`
`python test/test_quantization TestEqualizeFx.test_input_weight_equalization_weights_bias`
`python test/test_quantization TestEqualizeFx.test_input_activation_values`
`python test/test_quantization TestEqualizeFx.test_input_weight_equalization_graphs`

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

angelayi committed:
Update on "[quant] Input-Weight Equalization - tests"


Summary: Created the following unit-tests to check that our equalization algorithm is as expected:
- Check the equalization scales calculated and stored in the graph are as expected
- Check the scaled weights and biases are as expected
- Check that the min/max values in the quantization observers are as expected
- Check that the graphs with equalization are structured in the same way as graphs without equalization (except that equalized graphs have additional equalization scale and mul nodes) before and after quantization

Test Plan: `python test/test_quantization TestEqualizeFx.test_input_weight_equalization_equalization_scales`
`python test/test_quantization TestEqualizeFx.test_input_weight_equalization_weights_bias`
`python test/test_quantization TestEqualizeFx.test_input_activation_values`
`python test/test_quantization TestEqualizeFx.test_input_weight_equalization_graphs`

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

angelayi committed:
Update on "[quant] Input-Weight Equalization - tests"


Summary: Created the following unit-tests to check that our equalization algorithm is as expected:
- Check the equalization scales calculated and stored in the graph are as expected
- Check the scaled weights and biases are as expected
- Check that the min/max values in the quantization observers are as expected
- Check that the graphs with equalization are structured in the same way as graphs without equalization (except that equalized graphs have additional equalization scale and mul nodes) before and after quantization

Test Plan: `python test/test_quantization TestEqualizeFx.test_input_weight_equalization_equalization_scales`
`python test/test_quantization TestEqualizeFx.test_input_weight_equalization_weights_bias`
`python test/test_quantization TestEqualizeFx.test_input_activation_values`
`python test/test_quantization TestEqualizeFx.test_input_weight_equalization_graphs`

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

angelayi committed:
Update on "[quant] Input-Weight Equalization - tests"


Summary: Created the following unit-tests to check that our equalization algorithm is as expected:
- Check the equalization scales calculated and stored in the graph are as expected
- Check the scaled weights and biases are as expected
- Check that the min/max values in the quantization observers are as expected
- Check that the graphs with equalization are structured in the same way as graphs without equalization (except that equalized graphs have additional equalization scale and mul nodes) before and after quantization

Test Plan: `python test/test_quantization TestEqualizeFx.test_input_weight_equalization_equalization_scales`
`python test/test_quantization TestEqualizeFx.test_input_weight_equalization_weights_bias`
`python test/test_quantization TestEqualizeFx.test_input_activation_values`
`python test/test_quantization TestEqualizeFx.test_input_weight_equalization_graphs`

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

angelayi committed:
Update on "[quant] Input-Weight Equalization - tests"


Summary: Created the following unit-tests to check that our equalization algorithm is as expected:
- Check the equalization scales calculated and stored in the graph are as expected
- Check the scaled weights and biases are as expected
- Check that the min/max values in the quantization observers are as expected
- Check that the graphs with equalization are structured in the same way as graphs without equalization (except that equalized graphs have additional equalization scale and mul nodes) before and after quantization

Test Plan: `python test/test_quantization TestEqualizeFx.test_input_weight_equalization_equalization_scales`
`python test/test_quantization TestEqualizeFx.test_input_weight_equalization_weights_bias`
`python test/test_quantization TestEqualizeFx.test_input_activation_values`
`python test/test_quantization TestEqualizeFx.test_input_weight_equalization_graphs`

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

angelayi committed:
Update on "[quant] Input-Weight Equalization - tests"


Summary: Created the following unit-tests to check that our equalization algorithm is as expected:
- Check the equalization scales calculated and stored in the graph are as expected
- Check the scaled weights and biases are as expected
- Check that the min/max values in the quantization observers are as expected
- Check that the graphs with equalization are structured in the same way as graphs without equalization (except that equalized graphs have additional equalization scale and mul nodes) before and after quantization

Test Plan: `python test/test_quantization TestEqualizeFx.test_input_weight_equalization_equalization_scales`
`python test/test_quantization TestEqualizeFx.test_input_weight_equalization_weights_bias`
`python test/test_quantization TestEqualizeFx.test_input_activation_values`
`python test/test_quantization TestEqualizeFx.test_input_weight_equalization_graphs`

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

angelayi committed:
Update on "[quant] Input-Weight Equalization - tests"


Summary: Created the following unit-tests to check that our equalization algorithm is as expected:
- Check the equalization scales calculated and stored in the graph are as expected
- Check the scaled weights and biases are as expected
- Check that the min/max values in the quantization observers are as expected
- Check that the graphs with equalization are structured in the same way as graphs without equalization (except that equalized graphs have additional equalization scale and mul nodes) before and after quantization

Test Plan: `python test/test_quantization TestEqualizeFx.test_input_weight_equalization_equalization_scales`
`python test/test_quantization TestEqualizeFx.test_input_weight_equalization_weights_bias`
`python test/test_quantization TestEqualizeFx.test_input_activation_values`
`python test/test_quantization TestEqualizeFx.test_input_weight_equalization_graphs`

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

angelayi committed:
Update on "[quant] Input-Weight Equalization - tests"


Summary: Created the following unit-tests to check that our equalization algorithm is as expected:
- Check the equalization scales calculated and stored in the graph are as expected
- Check the scaled weights and biases are as expected
- Check that the min/max values in the quantization observers are as expected
- Check that the graphs with equalization are structured in the same way as graphs without equalization (except that equalized graphs have additional equalization scale and mul nodes) before and after quantization

Test Plan: `python test/test_quantization TestEqualizeFx.test_input_weight_equalization_equalization_scales`
`python test/test_quantization TestEqualizeFx.test_input_weight_equalization_weights_bias`
`python test/test_quantization TestEqualizeFx.test_input_activation_values`
`python test/test_quantization TestEqualizeFx.test_input_weight_equalization_graphs`

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

angelayi committed:
Update on "[quant] Input-Weight Equalization - tests"


Summary: Created the following unit-tests to check that our equalization algorithm is as expected:
- Check the equalization scales calculated and stored in the graph are as expected
- Check the scaled weights and biases are as expected
- Check that the min/max values in the quantization observers are as expected
- Check that the graphs with equalization are structured in the same way as graphs without equalization (except that equalized graphs have additional equalization scale and mul nodes) before and after quantization

Test Plan: `python test/test_quantization TestEqualizeFx.test_input_weight_equalization_equalization_scales`
`python test/test_quantization TestEqualizeFx.test_input_weight_equalization_weights_bias`
`python test/test_quantization TestEqualizeFx.test_input_activation_values`
`python test/test_quantization TestEqualizeFx.test_input_weight_equalization_graphs`

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

angelayi commented:
@angelayi has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29406942).

supriyar commented on test/quantization/fx/test_equalize_fx.py:
```diff
@@ -135,221 +158,373 @@ def test_input_weight_eq_observer(self, input_qdtype, input_qscheme, weight_qdty
+                    self.assertEqual(modules[str(node.target)].max_val, exp_max_val)
+                    weight_counter += 1
+
+    def check_orig_and_eq_graphs(self, orig_model, eq_model):
```
Remove this?

supriyar commented on test/quantization/fx/test_equalize_fx.py:
```diff
@@ -135,221 +158,373 @@ def test_input_weight_eq_observer(self, input_qdtype, input_qscheme, weight_qdty
+        observer's min/max values are as expected
+        """
+
+        tests = [SingleLayerLinearModel, SingleLayerFunctionalLinearModel]
```
Should there also be a TwoLinearLayer test here?

supriyar reviewed (approved):
Looks good, please address last couple comments

angelayi committed:
Update on "[quant] Input-Weight Equalization - tests"


Summary: Created the following unit-tests to check that our equalization algorithm is as expected:
- Check the equalization scales calculated and stored in the graph are as expected
- Check the scaled weights and biases are as expected
- Check that the min/max values in the quantization observers are as expected
- Check that the graphs with equalization are structured in the same way as graphs without equalization (except that equalized graphs have additional equalization scale and mul nodes) before and after quantization

Test Plan: `python test/test_quantization TestEqualizeFx.test_input_weight_equalization_equalization_scales`
`python test/test_quantization TestEqualizeFx.test_input_weight_equalization_weights_bias`
`python test/test_quantization TestEqualizeFx.test_input_activation_values`
`python test/test_quantization TestEqualizeFx.test_input_weight_equalization_graphs`

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29406942](https://our.internmc.facebook.com/intern/diff/D29406942)

[ghstack-poisoned]

angelayi committed:
Update on "[quant] Input-Weight Equalization - tests"


Summary: Created the following unit-tests to check that our equalization algorithm is as expected:
- Check the equalization scales calculated and stored in the graph are as expected
- Check the scaled weights and biases are as expected
- Check that the min/max values in the quantization observers are as expected
- Check that the graphs with equalization are structured in the same way as graphs without equalization (except that equalized graphs have additional equalization scale and mul nodes) before and after quantization

Test Plan: `python test/test_quantization TestEqualizeFx.test_input_weight_equalization_equalization_scales`
`python test/test_quantization TestEqualizeFx.test_input_weight_equalization_weights_bias`
`python test/test_quantization TestEqualizeFx.test_input_activation_values`
`python test/test_quantization TestEqualizeFx.test_input_weight_equalization_graphs`

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29406942](https://our.internmc.facebook.com/intern/diff/D29406942)

[ghstack-poisoned]

angelayi commented:
@angelayi has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29406942).
